Avoid server livelock by SystemDownHandler.

zalokhan opened the pull request:
## Overview

Description: Register a system down handler to prevent live lock conditions. Also, delegating the sequencer bootstrap task to the server for uniformity and non-blocking behaviour.

Why should this be merged: Prevent liveness issues in the cluster.

Related issue(s) (if applicable): #1374 


## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

StrikingThirteen commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -324,7 +326,7 @@ private void bootstrapPrimarySequencerServer() {
                 log.info("Bootstrapped sequencer server at epoch:{}", layout.getEpoch());
             }
-        } catch (InterruptedException | ExecutionException e) {
+        } catch (Exception e) {
```
Can't we come up with specific messages when we knew it was InterruptedException or ExecutionException? In other words, although, we are including the exception description with the message, is is possible that with this granularity we are losing some info about the problem?

Maithem replied:
Should not swallow the interrupt.

StrikingThirteen commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -69,6 +70,21 @@
+     * certain amount of time.
+     */
+    private final Runnable runtimeSystemDownHandler = () -> {
+        log.warn("ManagementServer: Runtime stalled. Invoking systemDownHandler.");
```
Would it be possible (and beneficial) to have more descriptive log message about trying to break out of live-lock rather than pointing to the mechanics of how we are doing it?

Also we are only logging and throwing an exception in the scope this runnable however the log implies of a handler being invoked which might be confusing.

zalokhan replied:
I'll add a descriptive log.

StrikingThirteen commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -69,6 +70,21 @@
+     * The number of retries to be made to execute any RPC request before the runtime gives up and
+     * invokes the systemDownHandler.
+     */
+    private static final int SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT = 50;
```
why 50? Maybe add the rationale in a comment

zalokhan replied:
I'll add the comment.

medhavidhawan commented on test/src/test/java/org/corfudb/runtime/view/ManagementViewTest.java:
```diff
@@ -1516,4 +1517,96 @@ public void queryClusterStatus() throws Exception {
+     * This triggers the systemDownHandler.
+     */
+    @Test
+    public void triggerSystemDownHandlerInDeadlock() throws Exception {
```
good test, maybe refactor the boiler plate.

corfu-repo commented:
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ManagementAgent.java#L19](https://github.com/CorfuDB/CorfuDB/blob/371628a4ccc2f685751c2ac10de00b825c0522f1/infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java#L19): Remove this unused import 'java.util.concurrent.RejectedExecutionException'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

Maithem commented:
Where you able to reproduce the bug?

medhavidhawan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -263,6 +281,42 @@ public synchronized void handleHealingDetectedMsg(CorfuPayloadMsg<DetectorMsg> m
+     * @param r   server router
+     */
+    @ServerHandler(type = CorfuMsgType.MANAGEMENT_SEQUENCER_BOOTSTRAP)
+    public synchronized void handleSequencerBootstrapTriggerMsg(CorfuPayloadMsg<Layout> msg,
```
Even then there could be slow arriving messages that could restart the FastLoader just after it was done. I think that needs to be fixed no matter what. You need to check if the sequencer has been bootstrapped before you start the FastLoader again.

medhavidhawan replied:
If that is the case you might get more benefit from queue. Other handler APIs if they are slow for whatever reason, will also not block each other.

corfu-repo commented:
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ManagementAgent.java#L19](https://github.com/CorfuDB/CorfuDB/blob/42b8af37868ac8fcf6ada25a96c782a2ee47bf94/infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java#L19): Remove this unused import 'java.util.concurrent.RejectedExecutionException'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

medhavidhawan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -274,23 +275,23 @@ private void initializationTask() {
+                    policyExecuteInterval,
+                    TimeUnit.MILLISECONDS);
+
         } catch (InterruptedException e) {
```
log something here.

medhavidhawan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -69,6 +70,28 @@
+        log.warn("ManagementServer: Runtime stalled. Invoking systemDownHandler after {} retries.",
+                SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT);
+        throw new StalledRuntimeException("Runtime stalled. Invoking systemDownHandler after "
+                + SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT + " retries.");
```
r/unsuccessful retries/tries

medhavidhawan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -69,6 +70,28 @@
+    };
+
+    /**
+     * The number of retries to be made to execute any RPC request before the runtime gives up and
```
r/retries/tries

medhavidhawan commented on runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java:
```diff
@@ -434,43 +439,58 @@ private long getMaxGlobalTail(Layout layout, Layout.LayoutSegment segment) {
-            log.info("Sequencer bootstrap successful.");
-        } else {
-            log.warn("Sequencer bootstrap failed. Already bootstrapped.");
+        boolean isLocked = recoverSequencerLock.tryLock();
```
r/isLocked/lockAcquired

medhavidhawan replied:
log if you did not get the lock

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -274,23 +275,24 @@ private void initializationTask() {
+                    policyExecuteInterval,
+                    TimeUnit.MILLISECONDS);
+
         } catch (InterruptedException e) {
```
Why do you differentiate between InterruptedException and Exception ? shouldn;t you throw an UnrecoverableCorfuError in both cases?

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -645,8 +647,11 @@ private void handleFailures(PollReport pollReport) {
+                        // If it fails, we detect this again and retry in the next polling cycle.
+                        getCorfuRuntime().getLayoutView().getRuntimeLayout(layout)
+                                .getManagementClient(getLocalEndpoint())
+                                .triggerSequencerBootstrap(layout).get();
```
If we don't care, then why do we even do a get on the future?

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java:
```diff
@@ -263,6 +288,41 @@ public synchronized void handleHealingDetectedMsg(CorfuPayloadMsg<DetectorMsg> m
+     * @param r   server router
+     */
+    @ServerHandler(type = CorfuMsgType.MANAGEMENT_SEQUENCER_BOOTSTRAP)
+    public void handleSequencerBootstrapTriggerMsg(CorfuPayloadMsg<Layout> msg,
```
I don't see a strong reason to push this on the server.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/exceptions/StalledRuntimeException.java:
```diff
@@ -0,0 +1,13 @@
+ */
+public class StalledRuntimeException extends RuntimeException {
+
+    public StalledRuntimeException(String message) {
```
Can probably use a generic exception.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java:
```diff
@@ -34,6 +35,10 @@ public LayoutManagementView(@NonNull CorfuRuntime runtime) {
 
     private volatile long prepareRank = 1L;
 
+    private final ReentrantLock recoverSequencerLock = new ReentrantLock();
```
Why do you need a lock, why can't you just use a volatile variable to signal ?

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -274,23 +275,24 @@ private void initializationTask() {
             }
 
             // Initiating periodic task to poll for failures.
-            try {
```
Please combine the logic of sequencer bootstrap for the case where it is a fresh start or recovery.

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -274,23 +275,24 @@ private void initializationTask() {
             // Initiating periodic task to poll for failures.
-            try {
-                localMetricsPollingService.scheduleAtFixedRate(
-                        this::updateLocalMetrics,
```
break up the runRecoveryReconfiguration so that the sequencer and epoch changes are separate.

Maithem commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/CorfuMsgType.java:
```diff
@@ -97,6 +97,7 @@
     HEARTBEAT_RESPONSE(76, new TypeToken<CorfuPayloadMsg<NodeView>>(){}, true),
     ORCHESTRATOR_REQUEST(77, new TypeToken<CorfuPayloadMsg<OrchestratorMsg>>() {}, true),
     ORCHESTRATOR_RESPONSE(78, new TypeToken<CorfuPayloadMsg<OrchestratorResponse>>() {}, true),
+    MANAGEMENT_SEQUENCER_BOOTSTRAP(79, new TypeToken<CorfuPayloadMsg<Layout>>() {}),
```
Move this to the agent and then you can remove the extra rpcs.

zalokhan commented:
@Maithem Addressed all comments except:
- Removing StreamLogFiles overloading of globalLogTail for no data and data on 0th Address.
- Moving to harness.

corfu-repo commented:
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

corfu-repo commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -800,6 +833,10 @@ public void shutdown() {
         try {
             initializationTaskThread.interrupt();
             initializationTaskThread.join(ServerContext.SHUTDOWN_TIMER.toMillis());
+        } catch (InterruptedException ie) {
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either re-interrupt this method or rethrow the "InterruptedException". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -226,6 +231,8 @@
+
         // Creating the initialization task thread.
         // This thread pool is utilized to dispatch one time recovery and sequencer bootstrap tasks.
         // One these tasks finish successfully, they initiate the detection tasks.
```
initializationTaskThread runs the sequencer recovery before scheduling the detector tasks, which was causing the problem in the first place, this dependency is till not fixed. Adding a handler to throw a cluster unavailable exception can throw an exception in initializationTaskThread before scheduling the detector tasks.

Maithem replied:
The initializationTaskThread will eventually call runLayoutReconfiguration which calls reconfigureSequencerServers. That method can block forever or throw an exception in either case the fault detectors wont be scheduled. That is one of the main problems and I don't see how this PR addresses that issue.

zalokhan replied:
The initializationTask does perform sequencer recovery but they are not blocking forever. They will now bail out with a clusterUnreachableException. These exceptions thrown by the sequener recovery task are logged and since they are best effort tasks, they do not affect the detector tasks.

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -226,6 +231,8 @@
                         .setNameFormat(serverContext.getThreadPrefix() + "LocalMetricsPolling")
                         .build());
 
+        this.sequencerRecoveryFuture = CompletableFuture.completedFuture(true);
```
you just care about this on init, can we remove these futures they are very confusing.

zalokhan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -110,12 +111,16 @@
      */
     @Getter
-    private volatile CompletableFuture<Boolean> sequencerBootstrappedFuture;
+    private volatile CompletableFuture<Boolean> recoveryBarrierFuture;
```
I plan to remove this with the removal of the bootstrapPrimarySequencer method as we discussed if that is fine. It is a bit messy currently but many of the tests also rely on this barrier and will need refactoring.

zalokhan commented:
We do solve the live-lock. Primarily by introducing system down handlers in the Runtime.
I agree the agent will be stalled until the recovery completes.
If you feel this is also unacceptable, we can delegate this task to the server as discussed earlier. Let me know what you feel would be the right approach design-wise.

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -110,12 +111,16 @@
+    private volatile CompletableFuture<Boolean> recoveryBarrierFuture;
+
+    /**
+     * Future which is reset every time a new task to bootstrap the sequencer is launched by the
```
Please simplify the nested while(true) loops.

zalokhan commented:
Will add logging for sequencer recovery. Resolves #1363

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -170,7 +175,7 @@
                 ? opts.get("--management-server").toString() : null;
 
-        sequencerBootstrappedFuture = new CompletableFuture<>();
+        recoveryBarrierFuture = new CompletableFuture<>();
```
Isn't this redundant ?

bootstrapPrimarySequencerServer will never execute before recoveryBarrierFuture is done (i.e. it will keep looping). Also, triggerSequencerBootstrap doesn't depend on its value, why have it ?

zalokhan replied:
I did attempt to remove it but was affecting a lot of tests. And I did not want to change any functionality. I just changed the name so that it is a little different and intuitive compared to the new CompletableFuture introduced - `sequencerRecoveryFuture`. I can do this with the cleanup we discussed.

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -238,6 +245,29 @@
+     * @return Completable future which completes when the task completes successfully or with a
+     * failure.
+     */
+    private CompletableFuture<Boolean> triggerSequencerBootstrap(@NonNull Layout layout) {
```
Should this be thread-safe? two threads can evaluate to sequencerRecoveryFuture.isDone()

zalokhan replied:
This is run in a single threaded manner. So we should not worry about thread safety. I will need to update it to an AtomicReference otherwise.

Maithem replied:
Ok. The logic is async, you can just use synchronized on the method.

zalokhan commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -247,50 +277,50 @@
-            if (!sequencerBootstrappedFuture.isDone()) {
+            // Sequencer bootstrap required if this is fresh startup.
+            if (!recoveryBarrierFuture.isDone()) {
                 bootstrapPrimarySequencerServer();
```
It is wrapped in an ExecutionException and caught in the method itself. It does not affect the scheduling of the detectors.

Maithem replied:
WrongEpochException derives from RuntimeException, how would catching ExecutionException address this?

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java:
```diff
@@ -434,43 +439,60 @@ private long getMaxGlobalTail(Layout layout, Layout.LayoutSegment segment) {
+            }
         } else {
-            log.warn("Sequencer bootstrap failed. Already bootstrapped.");
+            log.info("reconfigureSequencerServers: Sequencer reconfiguration already in progress.");
```
prefix with "reconfigureSequencerServers:"

zalokhan replied:
Done.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java:
```diff
@@ -434,43 +439,60 @@ private long getMaxGlobalTail(Layout layout, Layout.LayoutSegment segment) {
+                        || !originalLayout.getSequencers().get(0).equals(newLayout.getSequencers()
+                        .get(0))) {
+                    Layout.LayoutSegment latestSegment = newLayout.getSegments()
+                            .get(newLayout.getSegments().size() - 1);
```
You can actually fix the hole issue here, instead of only querying the tail query all logunits.

zalokhan replied:
I can but the new Runtimes will have the issue anyway. It would be cleaner to fix it during the stateTransfer.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java:
```diff
@@ -434,43 +439,60 @@ private long getMaxGlobalTail(Layout layout, Layout.LayoutSegment segment) {
+                // Reconfigure Primary Sequencer if required
+                if (forceReconfigure
+                        || !originalLayout.getSequencers().get(0).equals(newLayout.getSequencers()
+                        .get(0))) {
```
getPrimarySequencer

zalokhan replied:
Done.

Maithem commented on test/src/test/java/org/corfudb/runtime/view/ManagementViewTest.java:
```diff
@@ -1516,4 +1517,104 @@ public void queryClusterStatus() throws Exception {
+        semaphore.acquire(2);
+
+        //Release the semaphore on invocation.
+        Runnable systemDownHandler = () -> {
```
shouldn't this also test the actual handler? why redefine another one just for the test?

zalokhan replied:
You are right, I just wanted a way to know that this handler was invoked.
If I remove this, this test will become a black-box test and I will have to rely and assert on the fact that the detector is not stuck and handles some failures. Is that reasonable?

Maithem commented on test/src/test/java/org/corfudb/runtime/view/ManagementViewTest.java:
```diff
@@ -1516,4 +1517,104 @@ public void queryClusterStatus() throws Exception {
+        addServerRule(SERVERS.PORT_1, new TestRule().matches(m -> m.getMsgType().equals(CorfuMsgType.READ_RESPONSE)).drop());
+
+        Layout newLayout = new Layout(l);
+        newLayout.setEpoch(newLayout.getEpoch() + 1);
```
This doesn;t really test how the UnreachableClusterException is handled, which is the code path that your changing. I think we need a better test.

zalokhan commented:
@Maithem I have addressed some of your comments. I will take care of the rest of the changes during the cleanup and centralizing the FastObjectLoader on the Sequencer.

zalokhan commented:
@Maithem Fixed the test.

corfu-repo commented on infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java:
```diff
@@ -1,5 +1,6 @@
 package org.corfudb.infrastructure;
 
+import com.google.common.annotations.VisibleForTesting;
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.google.common.annotations.VisibleForTesting'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

corfu-repo commented:
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
